src/wp-includes/blocks/template-part.php

```diff
@@ -272,12 +272,12 @@ function build_template_part_block_variations() {
  * Registers the `core/template-part` block on the server.
  */
 function register_block_core_template_part() {
-	register_block_type_from_metadata(
-		__DIR__ . '/template-part',
-		array(
-			'render_callback' => 'render_block_core_template_part',
-			'variations'      => build_template_part_block_variations(),
-		)
+	$args = array(
+		'render_callback' => 'render_block_core_template_part',
 	);
+	if ( current_user_can( 'edit_theme_options' ) ) {
+			$args['variations'] = build_template_part_block_variations();
```

Comment:
I wonder if the capability check should be moved into the `build_template_part_block_variations` function itself and return an empty array, or `null` if false. I also wonder if this is even the appropriate capability check, since there might be valid use cases where someone needs to get a list of the registered variations even though they can't edit theme options directly (though, I don't have an idea what that use case might be at the moment).

Reply:
Yes @joemcgill , we can move it to `build_template_part_block_variations`. Give a thought about capability I was thinking of using `is_user_logged_in` instead. This way we will only restrict it to all the users that are logged in with out touching a specific capability.

Reply:
I don't mind the idea of using the capabilities API as long as we're sure we're identifying the correct cap needed to make use of these variations. Otherwise, even logged in users without any editing capabilities will get these variations on the front-end. 

Reading back through [the original thread](https://github.com/WordPress/gutenberg/issues/45601) about this in the Gutenberg repo, it seems like this is only used when editing templates, but @Mamaduka preferred a more declarative approach at the block API level, rather than needing to handle this on a block-by-block basis. That seems to be what @spacedmonkey was attempting to solve in https://github.com/WordPress/wordpress-develop/pull/4857. I tend to agree, and think we have time to be thoughtful about how to declaratively load variants only when needed, or make the variant property a callback that is only loaded when the variants are needed. What do you think?

Reply:
Thanks for the ping, @joemcgill!

I still think a declarative approach for lazy-loading would be best suited here. There are other blocks that generate variations on the server where the capabilities approach might not be applicable.

@spacedmonkey, I know you're taking a well-deserved break from contributing, but if you could share your notes for https://github.com/WordPress/wordpress-develop/pull/4857, maybe someone could pick it up again.

Reply:
@spacedmonkey and @Mamaduka, re:

> Side note, this change should be done in gutenberg, as this files are versioned there.

This PR, and the related Trac ticket were specifically created based on the conversation [the two of you had here](https://github.com/WordPress/gutenberg/issues/45601#issuecomment-1552441434). In which, @Mamaduka said that we needed these changes to be made in WP. Could one of you two clarify which changes you were referring to?

Reply:
@joemcgill, that’s for changes required to lazy-load variations property. We can’t override that from the plugin.